codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -165,8 +156,11 @@ public void onHttpClientUpgrade() throws Http2Exception {
         if (connection().isServer()) {
             throw connectionError(PROTOCOL_ERROR, "Client-side HTTP upgrade requested for a server");
         }
-        if (prefaceSent() || decoder.prefaceReceived()) {
-            throw connectionError(PROTOCOL_ERROR, "HTTP upgrade must occur before HTTP/2 preface is sent or received");
+        if (!prefaceSent()) {
+            throw connectionError(PROTOCOL_ERROR, "HTTP upgrade must occur after preface was sent");
```

Comment:
this seems backwards according to the spec, since you don't send the preface until you know the upgrade has succeeded.  can we add a note that this is an implementation detail in netty (maybe a comment?)

Reply:
Should this be INTERNAL instead? I agree it should be made more obvious to be a Netty-internal.

Reply:
yeah sounds good.